### PR TITLE
mumps: Migrate to homebrew-core

### PR DESCRIPTION
**Pull request by McNoggins**
Migrating as per Homebrew/homebrew-science#6331. Should pass audit

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/mumps.rb [hunk 1]
```diff
@@ -0,0 +1,201 @@
+class Mumps < Formula
+  desc "Parallel Sparse Direct Solver"
+  homepage "http://mumps-solver.org"
+  url "http://mumps.enseeiht.fr/MUMPS_5.1.2.tar.gz"
+  sha256 "eb345cda145da9aea01b851d17e54e7eef08e16bfa148100ac1f7f046cd42ae9"
+
+  depends_on :mpi => [:cc, :cxx, :f90, :recommended]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: please remove the `:recommended`

File: Formula/mumps.rb [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+  depends_on :fortran
+  depends_on "scalapack"
+  depends_on "metis"    => :optional if build.without? "mpi"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: please remove this line.

File: Formula/mumps.rb [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  depends_on "parmetis" => :optional if build.with? "mpi"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: please remove this line.

File: Formula/mumps.rb [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  depends_on "scotch5"  => :optional
```
Comment (ilovezfs) [inline Formula/mumps.rb]: please remove this line.

File: Formula/mumps.rb [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+  depends_on "scotch"   => :optional
```
Comment (ilovezfs) [inline Formula/mumps.rb]: please remove this line.

File: Formula/mumps.rb [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+  resource "mumps_simple" do
+    url "https://github.com/dpo/mumps_simple/archive/v0.4.tar.gz"
+    sha256 "87d1fc87eb04cfa1cba0ca0a18f051b348a93b0b2c2e97279b23994664ee437e"
+  end
+
+  def install
+    make_args = ["RANLIB=echo"]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: why is this needed?

File: Formula/mumps.rb [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+    shlibs_args = ["LIBEXT=.dylib",
+                   "AR=#{ENV["FC"]} -dynamiclib -Wl,-install_name -Wl,#{lib}/$(notdir $@) -undefined dynamic_lookup -o "]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: Why this needed? It looks like an upstream bug and/or total lack of macOS support.

File: Formula/mumps.rb [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+    make_args += ["OPTF=-O", "CDEFS=-DAdd_"]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: Why is this needed?

File: Formula/mumps.rb [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    orderingsf = "-Dpord"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: Why is this needed?

File: Formula/mumps.rb [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+    makefile = build.with?("mpi") ? "Makefile.G95.PAR" : "Makefile.G95.SEQ"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: mpi will be required so the conditional should be removed

File: Formula/mumps.rb [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+    cp "Make.inc/" + makefile, "Makefile.inc"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: Is this overwriting a file?

File: Formula/mumps.rb [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+
+    if build.with? "scotch5"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this block should be removed since scotch5 option will be gone

File: Formula/mumps.rb [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+      make_args += ["SCOTCHDIR=#{Formula["scotch5"].opt_prefix}",
+                    "ISCOTCH=-I#{Formula["scotch5"].opt_include}"]
+
+      if build.with? "mpi"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: if can be removed since it will always build with mpi

File: Formula/mumps.rb [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        scotch_libs = "LSCOTCH=-L$(SCOTCHDIR)/lib -lptesmumps -lptscotch -lptscotcherr"
+        scotch_libs += " -lptscotchparmetis" if build.with? "parmetis"
+        make_args << scotch_libs
+        orderingsf << " -Dptscotch"
+      else
```
Comment (ilovezfs) [inline Formula/mumps.rb]: else block should be removed

File: Formula/mumps.rb [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+        scotch_libs = "LSCOTCH=-L$(SCOTCHDIR) -lesmumps -lscotch -lscotcherr"
+        scotch_libs += " -lscotchmetis" if build.with? "metis"
+        make_args << scotch_libs
+        orderingsf << " -Dscotch"
+      end
+    elsif build.with? "scotch"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: elsif block should be removed

File: Formula/mumps.rb [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+      make_args += ["SCOTCHDIR=#{Formula["scotch"].opt_prefix}",
+                    "ISCOTCH=-I#{Formula["scotch"].opt_include}"]
+
+      if build.with? "mpi"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: if should be removed since it will always build with mpi

File: Formula/mumps.rb [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+        orderingsf << " -Dptscotch"
+      else
+        scotch_libs = "LSCOTCH=-L$(SCOTCHDIR) -lscotch -lscotcherr -lscotcherrexit"
+        scotch_libs += "-lscotchmetis" if build.with? "metis"
+        make_args << scotch_libs
+        orderingsf << " -Dscotch"
+      end
+    end
+
+    if build.with? "parmetis"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this whole block should be removed since parametis will be gone

File: Formula/mumps.rb [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+      make_args += ["LMETISDIR=#{Formula["parmetis"].opt_lib}",
+                    "IMETIS=#{Formula["parmetis"].opt_include}",
+                    "LMETIS=-L#{Formula["parmetis"].opt_lib} -lparmetis -L#{Formula["metis"].opt_lib} -lmetis"]
+      orderingsf << " -Dparmetis"
+    elsif build.with? "metis"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this whole block should be removed since metis will be gone

File: Formula/mumps.rb [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+      make_args += ["LMETISDIR=#{Formula["metis"].opt_lib}",
+                    "IMETIS=#{Formula["metis"].opt_include}",
+                    "LMETIS=-L#{Formula["metis"].opt_lib} -lmetis"]
+      orderingsf << " -Dmetis"
+    end
+
+    make_args << "ORDERINGSF=#{orderingsf}"
+
+    if build.with? "mpi"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: if should be removed since it will always build with mpi

File: Formula/mumps.rb [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+                    "SCALAP=-L#{Formula["scalapack"].opt_lib} -lscalapack",
+                    "INCPAR=", # Let MPI compilers fill in the blanks.
+                    "LIBPAR=$(SCALAP)"]
+    else
+      make_args += ["CC=#{ENV["CC"]} -fPIC",
+                    "FC=#{ENV["FC"]} -fPIC",
+                    "FL=#{ENV["FC"]} -fPIC"]
+    end
+
+    if build.with? "openblas"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: no openblas option so this should be removed

File: Formula/mumps.rb [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+      make_args << "LIBBLAS=-L#{Formula["openblas"].opt_lib} -lopenblas"
+    elsif build.with? "veclibfort"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this will always be true

File: Formula/mumps.rb [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+      make_args << "LIBBLAS=-L#{Formula["veclibfort"].opt_lib} -lvecLibFort"
+    else
+      make_args << "LIBBLAS=-lblas -llapack"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this will never be true since we'll always build with veclibfort

File: Formula/mumps.rb [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+    end
+
+    ENV.deparallelize # Build fails in parallel on Mavericks.
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this should be removed

File: Formula/mumps.rb [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+
+    system "make", "alllib", *(shlibs_args + make_args)
+
+    lib.install Dir["lib/*"]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: It needs to use `make install`

File: Formula/mumps.rb [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+    lib.install "libseq/libmpiseq.dylib" if build.without? "mpi"
+
+    # Build static libraries (e.g., for Dolfin)
+    system "make", "alllib", *make_args
+    (libexec/"lib").install Dir["lib/*.a"]
+    (libexec/"lib").install "libseq/libmpiseq.a" if build.without? "mpi"
+
+    inreplace "examples/Makefile" do |s|
```
Comment (ilovezfs) [inline Formula/mumps.rb]: why is this needed? we should be doing string replacements in the examples.

File: Formula/mumps.rb [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+      s.change_make_var! "libdir", lib
+    end
+
+    libexec.install "include"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this should be installed via `make install`

File: Formula/mumps.rb [hunk 27]
```diff
@@ -0,0 +1,201 @@
...
+    include.install_symlink Dir[libexec/"include/*"]
+    # The following .h files may conflict with others related to MPI
+    # in /usr/local/include. Do not symlink them.
+    (libexec/"include").install Dir["libseq/*.h"] if build.without? "mpi"
+
+    doc.install Dir["doc/*.pdf"]
+    pkgshare.install "examples"
+
+    prefix.install "Makefile.inc"  # For the record.
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this should be droped.

File: Formula/mumps.rb [hunk 28]
```diff
@@ -0,0 +1,201 @@
...
+    File.open(prefix/"make_args.txt", "w") do |f|
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this should be dropped.

File: Formula/mumps.rb [hunk 29]
```diff
@@ -0,0 +1,201 @@
...
+      f.puts(make_args.join(" "))  # Record options passed to make.
+    end
+
+    if build.with? "mpi"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this will always be true.

File: Formula/mumps.rb [hunk 30]
```diff
@@ -0,0 +1,201 @@
...
+        simple_args = ["CC=#{ENV["MPICC"]}", "prefix=#{prefix}", "mumps_prefix=#{prefix}",
+                       "scalapack_libdir=#{Formula["scalapack"].opt_lib}"]
+        if build.with? "scotch5"
+          simple_args += ["scotch_libdir=#{Formula["scotch5"].opt_lib}",
+                          "scotch_libs=-L$(scotch_libdir) -lptesmumps -lptscotch -lptscotcherr"]
+        elsif build.with? "scotch"
+          simple_args += ["scotch_libdir=#{Formula["scotch"].opt_lib}",
+                          "scotch_libs=-L$(scotch_libdir) -lptscotch -lptscotcherr -lscotch"]
+        end
+        if build.with? "openblas"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this will never be true so the block should be removed

File: Formula/mumps.rb [hunk 31]
```diff
@@ -0,0 +1,201 @@
...
+          simple_args += ["blas_libdir=#{Formula["openblas"].opt_lib}",
+                          "blas_libs=-L$(blas_libdir) -lopenblas"]
+        end
+        system "make", "SHELL=/bin/bash", *simple_args
+        lib.install "libmumps_simple.dylib"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: This should be installed with `make install`

File: Formula/mumps.rb [hunk 32]
```diff
@@ -0,0 +1,201 @@
...
+        include.install "mumps_simple.h"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: ditto

File: Formula/mumps.rb [hunk 33]
```diff
@@ -0,0 +1,201 @@
...
+      end
+    end
+  end
+
+  def caveats
```
Comment (ilovezfs) [inline Formula/mumps.rb]: The caveats should be removed.

File: Formula/mumps.rb [hunk 34]
```diff
@@ -0,0 +1,201 @@
...
+      EOS
+    end
+    s
+  end
+
+  test do
+    ENV.fortran
+    cp_r pkgshare/"examples", testpath
+    opts = ["-I#{opt_include}", "-L#{opt_lib}", "-lmumps_common", "-lpord"]
+    if Tab.for_name("mumps").with? "openblas"
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this will never be true

File: Formula/mumps.rb [hunk 35]
```diff
@@ -0,0 +1,201 @@
...
+      when building software that depends on MUMPS.
+      EOS
+    end
+    s
+  end
+
+  test do
+    ENV.fortran
+    cp_r pkgshare/"examples", testpath
+    opts = ["-I#{opt_include}", "-L#{opt_lib}", "-lmumps_common", "-lpord"]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: should use `include` and `lib` not `opt_include` and `opt_lib` for the formula itself

File: Formula/mumps.rb [hunk 36]
```diff
@@ -0,0 +1,201 @@
...
+    if Tab.for_name("mumps").with? "openblas"
+      opts << "-L#{Formula["openblas"].opt_lib}" << "-lopenblas"
+    else
+      opts << "-L#{Formula["veclibfort"].opt_lib}" << "-lvecLibFort"
+    end
+    if Tab.for_name("mumps").with?("mpi")
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this will always be true

File: Formula/mumps.rb [hunk 37]
```diff
@@ -0,0 +1,201 @@
...
+    ENV.deparallelize # Build fails in parallel on Mavericks.
+
+    system "make", "alllib", *(shlibs_args + make_args)
+
+    lib.install Dir["lib/*"]
+    lib.install "libseq/libmpiseq.dylib" if build.without? "mpi"
+
+    # Build static libraries (e.g., for Dolfin)
+    system "make", "alllib", *make_args
+    (libexec/"lib").install Dir["lib/*.a"]
```
Comment (ilovezfs) [inline Formula/mumps.rb]: this should be installed with `make install`

**Review by ilovezfs — CHANGES_REQUESTED** [review]

File: Formula/mumps.rb (hunk 7 above)
Comment (McNoggins) [reply Formula/mumps.rb]: These lines were in the original formula on homebrew-science, not quite sur why that's here. I'll try to clean it up a bit.

File: Formula/mumps.rb (hunk 1 above)
Comment (jslee02) [reply Formula/mumps.rb]: @ilovezfs Could you explain why `:recommended` needs to be removed? My best guess is Homebrew would like to remove optional dependencies as many as possible so that most of formula can be installed using the bottles. 

However, I would like to migrate `ipopt` from `homebrew-science`, and it requires `mumps` built without `mpi`. In this case, what would be the best practice to formulate `mumps`?:
1. Remove `mpi` dependency from `mumps`. One concern would be some `mpi` dependent formula might want `mumps` built with `mpi`.
2. Leave `mpi` is an optional dependency of `mumps`.
